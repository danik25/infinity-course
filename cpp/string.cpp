/****************************************************************
*                              String
* 						=====================
* File Name: string.cpp
* Related files: string.hpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#include <cstring> /* strlen */
#include <iostream> /* cout */
#include "string.hpp"

namespace hrd9
{
	/* constructor */
	String::String(const char* str)
	{
		size_t len = strlen(str);
		m_str = new char [len + 1];
		strcpy(m_str, str);
	}

	/* distructor */
	String::~String()
	{
		delete [](m_str);
	}

	/* copy constructor */
	String::String(const String& other) : m_str(init_new_str(other.m_str))
	{
		;
	}

	/* copy assignment operator */
	String& String :: operator=(const String& other)
	{
		delete[] m_str;
		m_str = init_new_str(other.m_str);

		return(*this);
	}

/*===================== non member functions =====================*/

/*===================== member functions =====================*/

	String& String::operator+=(const String& other)
	{
		size_t total_len = strlen(m_str) + strlen(other.m_str);

		String tmp = m_str;
		delete [] m_str;

		m_str = new char [total_len + 1];
		char * runner = m_str;

		strcpy(runner, tmp.m_str);
		strcat(runner, other.m_str);

		return(*this);
	}

	size_t String::length() const
	{
		return(strlen(m_str));
	}

	const char& String::operator[](size_t index) const
	{
		return(m_str[index]);
	}

	char& String::operator[](size_t index)
	{
		return(m_str[index]);
	}

	const char* String::to_c_str() const
	{
		return(m_str);
	}

/*===================== friend functions =====================*/

	const String operator+(const String& lhs, const String& rhs)
	{
		String ans = lhs;
		ans += rhs;

		return(ans);
	}

	bool operator==(const String& lhs, const String& rhs)
	{
		return(!strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator!=(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator>(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator<(const String& lhs, const String& rhs)
	{
		return(strcmp(rhs.m_str, lhs.m_str));
	}

	std::ostream& operator<<(std::ostream& os, const String& str)
	{
		os << str.m_str;
		return(os);
	}

/*===================== assistance functions =====================*/

	char* String::init_new_str(const char *str)
	{
		size_t len = strlen(str);
		char *new_str = new char [len + 1];
		strcpy(new_str, str);

		return(new_str);
	}
}