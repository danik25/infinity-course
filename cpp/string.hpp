/****************************************************************
*                              String
* 						=====================
* File Name: string.hpp
* Related files: string.cpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#ifndef ILHRD_STRING_HPP
#define ILHRD_STRING_HPP

#include <iosfwd> //ostream forward declarations

namespace hrd9
{

class String;
//non member functions:
const String operator+(const String& lhs, const String& rhs);
bool operator==(const String& lhs, const String& rhs);
bool operator!=(const String& lhs, const String& rhs);
bool operator>(const String& lhs, const String& rhs);
bool operator<(const String& lhs, const String& rhs);
std::ostream& operator<<(std::ostream& os, const String& str);
std::istream& operator>>(std::istream& is, String& str);

class String
{

public:
	// explicitly non explicit constructor
	String(const char* str = "");
	String(const String& other);
	String& operator=(const String& other);
	~String();

	String& operator+=(const String& other);
	size_t length() const;

	const char& operator[](size_t index) const; //read string[]
	char& operator[](size_t index);             //string[i] assignment

	const char* to_c_str() const;

private:
	friend bool operator==(const String& lhs, const String& rhs);
	friend bool operator!=(const String& lhs, const String& rhs);
	friend bool operator>(const String& lhs, const String& rhs);
	friend bool operator<(const String& lhs, const String& rhs);
	friend std::ostream& operator<<(std::ostream& os, const String& str);
	friend std::istream& operator>>(std::istream& is, String& str);

	char* init_new_str(const char *str);

	char* m_str;
};
} //namespace hrd9




#endif /* ILHRD_STRING_HPP */