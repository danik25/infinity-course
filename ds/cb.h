/****************************************************************
*                Data Structures - Circular buffer              *
* Related files: cb.c cb_test.c cb.h                            *
* #Version: V.01                                                *
* writer: Dani Benjamin                                         *
* reviewer: Ofer Tauber                                         *
* last update: 3.4.19                                           *
****************************************************************/ 
/*
* Circular buffers (also known as ring buffers) are fixed-size buffers 
* that work as if the memory is contiguous & circular in nature.
* As memory is generated and consumed, data does not need to be
* reshuffled - rather, the writer/reader pointers are adjusted. When data is added,
* the writer pointer advances. When data is consumed, the reader pointer advances.
* If you reach the end of the buffer, the pointers simply wrap around
* to the beginning.
*/  

/* API Design
 
	How users will interact with a circular buffer:

	* create the circular buffer container with a buffer and element sizes
	* destroy a circular buffer container
	* be able to add number of bytes to the buffer
	* be able to get number of bytes from the buffer
	* know whether the buffer is empty
	* know the current number of elements in the buffer
	* know the capacity of the buffer
*/

#ifndef _CB_H_
#define _CB_H_

#include <stddef.h> /* size_t */

typedef struct circular_buffer cb_t;

/*	Creates the circular buffer structure and returns a circular buffer handle.
 *	Parameters:	capacity in bytes
 *
 *	Return: circular buffer handle
 *
 */
cb_t *CBCreate(size_t capacity);


/* 	Free a circular buffer structure and data buffer.
 *
 *	Parameters:	buffer - pointer to circular buffer
 */
void CBDestroy(cb_t *buffer);


/* Writes <bytes_num> from <src> to <buffer>
 * rejects new data if the buffer is full
 * Returns number of bytes written on success,
 * 0 if buffer is full.
 *
 * Parameters:	buffer - pointer to circular buffer
 *				src - pointer of data source
 * 				bytes_num - number of bytes to write
 *
 * Return value: number of bytes actualy written
 *				 or 0 if buffer is full.
 *
 * Time complexity O(bytes_num).
 */
size_t CBWrite(cb_t *buffer, const void *src, size_t bytes_num);


/* Reads <bytes_num> from <buffer> to <dest>
 * Returns number of bytes actualy readed,
 * 0 if buffer is empty.

 * Parameters:	buffer - pointer to circular buffer
 *				dest - pointer of destination data to be readed
 * 				bytes_num - number of bytes to read
 *
 * Return value: number of bytes actualy readed
 				 or 0 if buffer is empty.

 * Time complexity O(bytes_num).
 */
size_t CBRead(cb_t *buffer, void *dest, size_t bytes_num);


/* Checks if buffer is empty
 *
 * Parameters:	buffer - pointer to circular buffer
 *
 * Return value: 1 if buffer is empty, 0 otherwise.
 */
int CBIsEmpty(cb_t *buffer);


/*	Returns available space of the buffer.
 	Parameters:	buffer - pointer to circular buffer
	
	Return: number of available bytes
 */
size_t CBFreeSpace(cb_t *buffer);


/*	Returns total capacity of buffer.
 	Parameters:	buffer - pointer to circular buffer
	
	Return: capacity of circular buffer in bytes
 */
size_t CBCapacity(cb_t *buffer);


#endif /* _CB_H_ */
