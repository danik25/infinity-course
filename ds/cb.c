#include <stddef.h> /* size_t */
#include <assert.h> /* assert */
#include <stdlib.h> /* malloc */
#include <string.h> /* memmove */
#include "cb.h"
#include <errno.h> /* errno, perror */
#define BYTE_SIZE 8

struct circular_buffer
{
	void *array;			/*	circular buffer handle	*/ 
	void *reader;			/*	next element to read 	*/
	void *writer;			/*	next element to write	*/
	size_t size;			/*	number of elements		*/ 
	size_t capacity;		/*	capacity of buffer		*/
};

cb_t *CBCreate(size_t capacity)
{
    cb_t *circ = NULL;
    if (0 == capacity)
    {
        return NULL;
    }
    circ = (cb_t*) malloc(sizeof(cb_t));
    circ->array = (cb_t*) malloc(capacity);
    circ->capacity = capacity;
    circ->size = 0;
    circ->reader = circ->array;
    circ->writer = circ->array;
    
    if (NULL == circ)
    {
    	return NULL;
    }
    if (NULL == circ->array)
    {
        free(circ);
    	return NULL;
    }
    return circ;
}

void CBDestroy(cb_t *buffer)
{
    assert(buffer);
    assert(buffer->array);
    free(buffer->array);
    buffer->array = NULL;
    
    free(buffer);
    buffer = NULL;
}

size_t CBWrite(cb_t *buffer, const void *src, size_t bytes_num)
{
	size_t cnt = 0;
	assert(buffer);
	assert(src);
	while(bytes_num > 0 && buffer->size < buffer->capacity) 
    {     
        memmove((char*)buffer->writer, src, 1);
        ++cnt;
        if((((size_t)buffer->writer) % 
                        ((size_t)buffer->array)) >= buffer->capacity - 1)
        {
            buffer->writer = buffer->array;
        }
        else
        {
        	buffer->writer = ((char*)buffer->writer + 1);
        }
        ++buffer->size;
        src = (char*)src + 1;
        --bytes_num;
    }
    if(bytes_num)
    {
        errno = EOVERFLOW;
    }
    return cnt;
}
size_t CBRead(cb_t *buffer, void *dest, size_t bytes_num)
{
	size_t cnt = 0;
	assert(buffer);
	assert(dest);
	while(bytes_num > 0 && buffer->size > 0 && *(char*)buffer->reader)
    {     
        *(char*)dest = *(char*)buffer->reader;
        ++cnt;
        if((((size_t)buffer->reader) % 
                        ((size_t)buffer->array)) >= buffer->capacity - 1)
        {
            buffer->reader = buffer->array;
        }
        else
        {
        	buffer->reader = (char*)buffer->reader + 1;
        }
        --buffer->size;
        --bytes_num;
        dest = (char*)dest + 1;
    }
    if(bytes_num)
    {
    	errno = ENODATA;
    }
    return cnt;
}

int CBIsEmpty(cb_t *buffer)
{
    assert(buffer);
    if(0 == buffer->size)
    {
        return 1; /*buffer is empty*/
    }
    return 0;
}

size_t CBFreeSpace(cb_t *buffer)
{
    assert(buffer);
    return buffer->capacity - buffer->size;
}

size_t CBCapacity(cb_t *buffer)
{
    assert(buffer);
    return buffer->capacity;
}






