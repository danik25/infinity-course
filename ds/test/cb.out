/home/xyz/git/dani-benjamin/ds/cb.out